core/sql/parser/sqlparser.y

```diff
@@ -7001,6 +7036,46 @@ rel_subquery_and_as_clause : rel_subquery as_clause
                                $$ = $2;
                              }
 
+with_clause_list : with_clause
+                     {  $$ = $1 ; }
+
+                   | with_clause_list ',' correlation_name TOK_AS rel_subquery 
```

Comment:
One way would be:

with_clause : WITH with_clause_elements

with_clause_elements : with_clause element
                                  |
                                     with_clause_elements with_clause_element

with_clause_element : correlation_name TOK_AS rel_subquery

By the way, I notice the way this is coded, the "with_clause" non-terminal is a single correlation_name TOK_AS rel_subquery element. It seems more natural (to me anyway) to think of the with_clause as being everything from the WITH keyword through the entire list of subqueries.